File: note-python/graphs/spanning-tree/kruskal-algorithm.py
import sys


def getFile(fileName):
    path = sys.argv[0].split("/")
    path.pop()
    filePath = "/".join(path)
    file = open(filePath + "/" + fileName, "r")
    return file


def find_parent(parent, n):
    if parent[n] != n:
        parent[n] = find_parent(parent, parent[n])
    return parent[n]


def union_parent(parent, a, b):
    a = find_parent(parent, a)
    b = find_parent(parent, b)

    if a > b:
        parent[a] = b
    else:
        parent[b] = a


file = getFile("input")

V, E = map(int, file.readline().rstrip().split())
parent = [0] * (V + 1)
edges = []
answer = 0

for v in range(1, V + 1):
    parent[v] = v

for _ in range(E):
    start, end, cost = list(map(int, file.readline().rstrip().split()))
    edges.append((start, end, cost))

edges.sort(key=lambda x: x[2])

# 가장 비용이 낮은 edge 부터 순회하면서 cycle 이 없으면 union
for edge in edges:
    start, end, cost = edge

    if find_parent(parent, start) == find_parent(parent, end):
        continue
    union_parent(parent, start, end)
    answer += cost

print(answer)

file.close()

File: leetcode/note/kruskal-algorithm.js
function findParent(parent, n) {
  if (parent[n] !== n) {
    parent[n] = findParent(parent, parent[n]);
  }
  return parent[n];
}

function unionParent(parent, a, b) {
  a = findParent(parent, a);
  b = findParent(parent, b);

  if (a > b) {
    parent[a] = b;
  } else {
    parent[b] = a;
  }
}

let minCost = 0;
const V = 7;
const E = 9;
const edges = [
  [1, 2, 29],
  [1, 5, 75],
  [2, 3, 35],
  [2, 6, 34],
  [3, 4, 7],
  [4, 6, 23],
  [4, 7, 13],
  [5, 6, 53],
  [6, 7, 25],
];
const parent = Array(V + 1).fill(0);
edges.sort((a, b) => a[2] - b[2]);

for (let i = 1; i < V + 1; i++) {
  parent[i] = i;
}

for (const edge of edges) {
  const [a, b, cost] = edge;

  if (findParent(parent, a) === findParent(parent, b)) {
    continue;
  }
  unionParent(parent, a, b);
  minCost += cost;
}

console.log(minCost);
